src/util/metadata/side_metadata/global.rs

```diff
@@ -946,6 +1075,194 @@ impl SideMetadataSpec {
+        }
+
+        // Figure out the start and end data address.
+        let start_addr = data_addr.saturating_sub(search_limit_bytes) + 1usize;
+        let end_addr = data_addr;
+
+        // Then figure out the start and end metadata address and bits.
+        let start_meta_addr = address_to_contiguous_meta_address(self, start_addr);
+        let start_meta_shift = meta_byte_lshift(self, start_addr);
+        let end_meta_addr = address_to_contiguous_meta_address(self, end_addr);
+        let end_meta_shift = meta_byte_lshift(self, end_addr);
+
+        // The result will be set by one of the following closures.
+        // Use Cell so it doesn't need to be mutably borrowed by the two closures which Rust will complain.
+        let res = std::cell::Cell::new(None);
+
+        // Check and find the last bit that is set. We try load words where possible, and fall back to load bytes.
+        let check_bytes_backwards = |start: Address, end: Address| -> bool {
+            use crate::util::constants::BYTES_IN_ADDRESS;
+            unsafe {
+                let mut cur = end;
+                while cur > start {
+                    // If we can check the whole word, set step to word size. Otherwise, the step is 1 (byte) and we check byte.
+                    let step = if cur.is_aligned_to(BYTES_IN_ADDRESS)
+                        && cur.align_down(BYTES_IN_ADDRESS) >= start
+                    {
+                        BYTES_IN_ADDRESS
+                    } else {
+                        1
+                    };
+                    // Move to the address so we can load from it
+                    cur -= step;
+
+                    // We have to always check if the metadata address is mapped or not.
+                    if !cur.is_mapped() {
+                        // Return true for early abort. If we see an unmapped metadata address, there is no point to keep checking.
+                        return true;
+                    }
+
+                    if step == BYTES_IN_ADDRESS {
+                        // Load and check a usize word
+                        let value = cur.load::<usize>();
+                        if value != 0 {
+                            // Find the exact non-zero byte within the usize using bitwise operations
+                            let byte_offset = (value.trailing_zeros() / 8) as usize;
+                            let byte_addr = cur + byte_offset;
+                            let byte_value = (value >> (byte_offset * 8)) & 0xFF;
+                            if byte_value != 0 {
+                                let bit = byte_value.trailing_zeros();
+                                debug_assert!(bit < 8);
+                                res.set(Some(contiguous_meta_address_to_address(
+                                    self, byte_addr, bit as u8,
+                                )));
+                                return true;
```

Comment:
This looks identical to line 1213 to 1218.  This means if we can extract the current lambda into a function that returns the last non-zero byte, the code paths can be reused.